### PR TITLE
Split tip-pools by lazyness

### PR DESCRIPTION
**Pull request by muXxer**
This PR is a draft for my "tangle split" idea to counter attack Thoralf's below max depth attack. 
Lets call it "split an conquer" :carlos:

### CODE REVIEW

File: pkg/model/coordinator/coordinator.go [hunk 1]
```diff
@@ -163,8 +157,10 @@ func (coo *Coordinator) InitState(bootstrap bool, startIndex milestone.Index) er
 			return fmt.Errorf("previous milestone does not match latest milestone in database: %d != %d", startIndex-1, latestMilestoneFromDatabase)
 		}
 
-		// if we bootstrap a network, NullHash has to be set as a solid entry point
-		tangle.SolidEntryPointsAdd(hornet.NullHashBytes, startIndex)
+		if startIndex == 1 {
```
Comment (alexsporn) [inline pkg/model/coordinator/coordinator.go:160]: I still think this is wrong. You can bootstrap the network at any index. e.g. after someone decides to do a global snapshot add add a gap in the milestones

File: pkg/model/coordinator/coordinator.go [hunk 2]
```diff
@@ -308,68 +252,78 @@ func (coo *Coordinator) createAndSendMilestone(trunkHash hornet.Hash, branchHash
...
-	return nil
+	return coo.state.LatestMilestoneHash, nil
 }
 
 // IssueCheckpoint tries to create and send a "checkpoint" to the network.
 // a checkpoint can contain multiple chained transactions to reference big parts of the unconfirmed cone.
 // this is done to keep the confirmation rate as high as possible, even if there is an attack ongoing.
 // new checkpoints always reference the last checkpoint or the last milestone if it is the first checkpoint after a new milestone.
-func (coo *Coordinator) IssueCheckpoint() error {
+func (coo *Coordinator) IssueCheckpoint(checkpointName string, checkpointIndex int, lastCheckpointHash hornet.Hash, tips hornet.Hashes) (hornet.Hash, error) {
```
Comment (alexsporn) [inline pkg/model/coordinator/coordinator.go]: checkpointName?

File: pkg/model/coordinator/coordinator.go [hunk 3]
```diff
@@ -308,68 +252,78 @@ func (coo *Coordinator) createAndSendMilestone(trunkHash hornet.Hash, branchHash
...
+		lastCheckpointHash = hornet.HashFromHashTrytes(b[0].Hash)
+
+		coo.Events.IssuedCheckpointTransaction.Trigger(checkpointName, checkpointIndex, i, len(tips), lastCheckpointHash)
+	}
+
+	return lastCheckpointHash, nil
 }
 
 // IssueMilestone creates the next milestone.
 // a new checkpoint is created right in front of the milestone to raise confirmation rate.
```
Comment (alexsporn) [inline pkg/model/coordinator/coordinator.go]: This is no longer valid

File: plugins/coordinator/plugin.go [hunk 4]
```diff
@@ -141,36 +166,135 @@ func run(plugin *node.Plugin) {
...
+				}
+
+				milestoneHash, err, criticalErr := coo.IssueMilestone(lastMilestoneHash, lastCheckpointHashNonLazy)
 				if criticalErr != nil {
 					log.Panic(criticalErr)
 				}
 				if err != nil {
+					if err == tangle.ErrNodeNotSynced {
+						// Coordinator is not synchronized, trigger the solidifier manually
+						tangleplugin.TriggerSolidifier()
```
Comment (alexsporn) [inline plugins/coordinator/plugin.go:217]: 🙈

File: plugins/coordinator/plugin.go [hunk 5]
```diff
@@ -141,36 +166,135 @@ func run(plugin *node.Plugin) {
...
+					}
 					log.Warn(err)
 				}
 
+				// remember the last milestone hash
+				lastMilestoneHash = milestoneHash
```
Comment (alexsporn) [inline plugins/coordinator/plugin.go:224]: if you had an error when issuing the milestone, which value is inside milestoneHash?
Maybe break inside the error check?

File: plugins/coordinator/plugin.go [hunk 6]
```diff
@@ -241,14 +383,36 @@ func configureEvents() {
...
+			score := tipselect.CalculateScore(bndl.GetTailHash(), tangle.GetSolidMilestoneIndex(), maxDeltaTxYoungestRootSnapshotIndexToLSMI, belowMaxDepth, maxDeltaTxApproveesOldestRootSnapshotIndexToLSMI)
+			switch score {
+			case tipselect.ScoreLazy:
+				// we ignore tips that are lazy.
+				return
+
+			case tipselect.ScoreSemiLazy:
+				// add semi-lazy tips to the semi-lazy heaviest branch selector
+				if trackedTailsCount := selectorSemiLazy.OnNewSolidBundle(bndl); trackedTailsCount >= maxTrackedTails {
+					log.Debugf("Coordinator Tipselector: trackedTailsCount: %d", trackedTailsCount)
```
Comment (alexsporn) [inline plugins/coordinator/plugin.go]: add "semi-lazy" to log?

File: plugins/coordinator/plugin.go (hunk 4 above)
Comment (muXxer) [reply plugins/coordinator/plugin.go:217]: :speak_no_evil:

File: pkg/model/coordinator/coordinator.go (hunk 1 above)
Comment (muXxer) [reply pkg/model/coordinator/coordinator.go:160]: But then I don't want to add NullHashBytes as a solid entrypoint. COO should reference the last issued milestone even with gaps.... (but why add gaps at all :scream: )

**Review by alexsporn — APPROVED** [review]

**Pull request merged**